ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/recovery/PeriodicRecovery.java

```diff
@@ -186,31 +188,81 @@ public void shutdown (boolean async)
    }
 
     /**
-     * make all scanning operations suspend.
+     * <p>Make all scanning operations suspend.
+     * <p>This switches the recovery operation mode to <code>SUSPENDED</code>.
+     * Any attempt to start a new scan either by an ad hoc threads or by the periodic
+     * recovery thread will suspend its thread until the mode changes.
+     * If a scan is in progress when this method is called it will complete its scan
+     * without suspending.
+     * <p>Note that this method is also influenced by
+     * {@link com.arjuna.ats.arjuna.common.RecoveryEnvironmentBean#setWaitForRecovery}.
+     * In case <code>{@link RecoveryEnvironmentBean#isWaitForRecovery()} == true</code>,
+     * it is <b>very important</b> that, before invoking this method, the following
+     * pre-conditions are met:
+     * * The Transaction System gets disabled (i.e. {@link TxControl#disable()})
+     * (i.e. no new transactions will be created)
+     * * {@link com.arjuna.ats.arjuna.coordinator.TransactionReaper} finishes monitoring all in-flight
+     * transactions
+     * * All transactions without a timeout are beyond the prepare phase of the 2PC protocol;
+     * to check when this condition is verified, {@link com.arjuna.ats.arjuna.coordinator.ActionManager}
+     * can be employed
      *
-     * This switches the recovery operation mode to SUSPENDED. Any attempt to start a new scan either by an ad hoc
-     * threads or by the periodic recovery thread will suspend its thread until the mode changes. If a scan is in
-     * progress when this method is called it will complete its scan without suspending.
-     *
-     * @param async false if the calling thread should wait for any in-progress scan to complete before returning
-     * @return the previous mode before attempting the suspend
+     * @param async false if the calling thread should wait for any in-progress scan to complete before returning.
+     * In case <code>{@link RecoveryEnvironmentBean#isWaitForRecovery()} == true</code>, this parameter is
+     * override.
+     * @return the previous mode before attempting the suspension
      */
-
    public Mode suspendScan (boolean async)
    {
        synchronized (_stateLock)
        {
-           // only switch and kick everyone if we are currently ENABLED
            Mode currentMode = getMode();
 
+           /*
+            * isWaitForRecovery checks whether suspension should delay until RecoveryModule
+            * implementations recover all their transactions.
+            */
+           if (currentMode == Mode.ENABLED && recoveryPropertyManager.getRecoveryEnvironmentBean().isWaitForRecovery()) {
+
+               /*
+                * At least, one clear scan (after the Transaction Manager has been shut down)
+                * should be completed to make sure that there are no transactions to recover.
+                */
+               doScanningWait();
+               doWork();
```

Comment:
Can't we come up with a design that plays nicely with the existing run loop rather than have two threads doing the work since that would reduce the probability of introducing bugs into this piece of well crafted code. It's a shame we can't ask Andrew to do the review.

Reply:
ok, that's fair so I will resolve this conversation.